libavfilter/sf_text2graphicsub.c

```diff
@@ -0,0 +1,492 @@
+/*
+ * Copyright (c) 2021 tcoza
+ *
+ * This file is part of FFmpeg.
+ *
+ * FFmpeg is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 2.1 of the License, or (at your option) any later version.
+ *
+ * FFmpeg is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with FFmpeg; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA
+ */
+
+/**
+ * @file
+ * convert text subtitles to bitmap subtitles filter
+ */
+
+#include <ass/ass.h>
+#include "libavutil/avstring.h"
+#include "libavutil/ass_split_internal.h"
+
+#include "internal.h"
+#include "avfilter.h"
+#include "drawutils.h"
+#include "libavutil/opt.h"
+#include "libavutil/buffer.h"
+#include "libavutil/internal.h"
+#include "libavformat/avformat.h"
+#include "libavcodec/elbg.h"
+#include "libavutil/lfg.h"
+
+typedef struct PalettizeContext {
+    int *codeword, *codebook;
+    int *codeword_closest_codebook_idxs;
+    int r_idx, g_idx, b_idx, a_idx;
+    struct ELBGContext *elbg;
+    AVLFG lfg;
+} PalettizeContext;
+
+typedef struct ASS_Context {
+    ASS_Library *library;
+    ASS_Renderer *renderer;
+    ASS_Track *track;
+    ASSSplitContext *ass_split_context;
+    PalettizeContext *palettize_context;
+    FFDrawContext draw_context;
+    struct { int width; int height; } size;
+    int num_colors;
+    char *filename, *fontsdir;
+} ASS_Context;
+
+static void ass_log(int ass_level, const char *fmt, va_list args, void *ctx);
+static void ass_image_to_area_palletization(ASS_Context *context, ASS_Image *image, AVSubtitleArea *area);
+static void init_palettizecontext(PalettizeContext **palettize_context);
+static void free_palettizecontext(PalettizeContext **palettize_context);
+static int stream_is_font(AVStream * st);
```

Comment:
What do you mean without need? They are all needed there.

Reply:
> Please just order the functions accordingly.

They are not needed. This is your filter code without forward declarations:

```
/*
 * Copyright (c) 2021 tcoza
 *
 * This file is part of FFmpeg.
 *
 * FFmpeg is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 2.1 of the License, or (at your option) any later version.
 *
 * FFmpeg is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public
 * License along with FFmpeg; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA
 */

/**
 * @file
 * convert text subtitles to bitmap subtitles filter
 */

#include <ass/ass.h>
#include "libavutil/avstring.h"
#include "libavutil/ass_split_internal.h"

#include "internal.h"
#include "avfilter.h"
#include "drawutils.h"
#include "libavutil/opt.h"
#include "libavutil/buffer.h"
#include "libavutil/internal.h"
#include "libavformat/avformat.h"
#include "libavcodec/elbg.h"
#include "libavutil/lfg.h"

typedef struct PalettizeContext {
    int *codeword, *codebook;
    int *codeword_closest_codebook_idxs;
    int r_idx, g_idx, b_idx, a_idx;
    struct ELBGContext *elbg;
    AVLFG lfg;
} PalettizeContext;

typedef struct ASS_Context {
    ASS_Library *library;
    ASS_Renderer *renderer;
    ASS_Track *track;
    ASSSplitContext *ass_split_context;
    PalettizeContext *palettize_context;
    FFDrawContext draw_context;
    struct { int width; int height; } size;
    int num_colors;
    char *filename, *fontsdir;
} ASS_Context;

/* libass supports a log level ranging from 0 to 7 */
static const int ass_libavfilter_log_level_map[] = {
    [0] = AV_LOG_FATAL,     /* MSGL_FATAL */
    [1] = AV_LOG_ERROR,     /* MSGL_ERR */
    [2] = AV_LOG_WARNING,   /* MSGL_WARN */
    [3] = AV_LOG_WARNING,   /* <undefined> */
    [4] = AV_LOG_INFO,      /* MSGL_INFO */
    [5] = AV_LOG_INFO,      /* <undefined> */
    [6] = AV_LOG_VERBOSE,   /* MSGL_V */
    [7] = AV_LOG_DEBUG,     /* MSGL_DBG2 */
};

static void ass_log(int ass_level, const char *fmt, va_list args, void *ctx)
{
    const int ass_level_clip = av_clip(ass_level, 0,
        FF_ARRAY_ELEMS(ass_libavfilter_log_level_map) - 1);
    const int level = ass_libavfilter_log_level_map[ass_level_clip];
    av_vlog(NULL, level, fmt, args);
    av_log(NULL, level, "\n");
}

static void set_area_bounds(ASS_Image *image, AVSubtitleArea *area)
{
    ASS_Image *image_o = image;
    area->x = image->dst_x + 1; area->w = -1;
    area->y = image->dst_y + 1; area->h = -1;
    for (image = image_o; image != NULL; image = image->next)
    {
        // Set image position to encompass all images
        if (image->dst_x < area->x) area->x = image->dst_x;
        if (image->dst_y < area->y) area->y = image->dst_y;
    }
    for (image = image_o; image != NULL; image = image->next)
    {
        // Set image size to encompass all images
        if (image->dst_x + image->w > area->x + area->w)
            area->w = image->dst_x + image->w - area->x;
        if (image->dst_y + image->h > area->y + area->h)
            area->h = image->dst_y + image->h - area->y;
    }
    area->linesize[0] = area->w;            // How it is
}

#define ALPHA_THRESHOLD 0b10000000
#define ALPHA_MAX ((1 << (sizeof(uint8_t) << 3)) - 1)
static void palettize_image(PalettizeContext *const s,
                            const int w, const int h,
                            uint8_t *src_data, const int src_linesize,
                            uint8_t *dst_data, const int dst_linesize,
                            uint32_t *dst_pal, const int num_output_colors);

/* libass stores an RGBA color in the format RRGGBBTT, where TT is the transparency level */
#define AR(c)  (((c)>>24)&0xFF)
#define AG(c)  (((c)>>16)&0xFF)
#define AB(c)  (((c)>> 8)&0xFF)
#define AA(c)  ((0xFF-(c))&0xFF)
#define TO_AREA_COLOR(c) (AB(c)|(AG(c)<<8)|(AR(c)<<16)|(AA(c)<<24))

static void ass_image_to_area_palletization(ASS_Context *context, ASS_Image *image, AVSubtitleArea *area)
{
    unsigned nb_pixels;
    size_t image_rgba_size;
    uint8_t *image_rgba;

    set_area_bounds(image, area);

    // Create rgba image
    nb_pixels = area->w * area->h;
    image_rgba_size = nb_pixels * 4 * sizeof(uint8_t);
    image_rgba = (uint8_t *)av_mallocz(image_rgba_size);
    for (; image != NULL; image = image->next)
    {
        uint8_t rgba_color[] = {AR(image->color), AG(image->color), AB(image->color), AA(image->color)};
        int linesize = area->w * 4;
        FFDrawColor color;
        ff_draw_color(&context->draw_context, &color, rgba_color);
        ff_blend_mask(&context->draw_context, &color,
                      &image_rgba, &linesize, area->w, area->h,
                      image->bitmap, image->stride, image->w, image->h,
                      3, 0, image->dst_x - area->x, image->dst_y - area->y);
    }

    area->nb_colors = context->num_colors;
    area->buf[0] = av_buffer_alloc(nb_pixels * sizeof(uint8_t));

    palettize_image(context->palettize_context,
                    area->w, area->h,
                    image_rgba, area->w * 4,
                    area->buf[0]->data, area->w,
                    &area->pal[0], context->num_colors);

    // Clean up
    av_free(image_rgba);
}

static void init_palettizecontext(PalettizeContext **palettizecontext)
{
    uint8_t rgba_map[4];
    PalettizeContext *context = (PalettizeContext *)av_malloc(sizeof(PalettizeContext));
    context->codebook = NULL;
    context->codeword = NULL;
    context->codeword_closest_codebook_idxs = NULL;
    context->elbg = NULL;
    av_lfg_init(&context->lfg, 0xACBADF);
    ff_fill_rgba_map(&rgba_map[0], AV_PIX_FMT_RGB32);
    context->r_idx = rgba_map[0]; // R
    context->g_idx = rgba_map[1]; // G
    context->b_idx = rgba_map[2]; // B
    context->a_idx = rgba_map[3]; // A
    *palettizecontext = context;
}

static void free_palettizecontext(PalettizeContext **palettizecontext)
{
    PalettizeContext *context = *palettizecontext;
    av_freep(&context->codebook);
    av_freep(&context->codeword);
    av_freep(&context->codeword_closest_codebook_idxs);
    avpriv_elbg_free(&context->elbg);
    av_free(context);
    *palettizecontext = NULL;
}

static const char * const font_mimetypes[] = {
        "font/ttf",
        "font/otf",
        "font/sfnt",
        "font/woff",
        "font/woff2",
        "application/font-sfnt",
        "application/font-woff",
        "application/x-truetype-font",
        "application/vnd.ms-opentype",
        "application/x-font-ttf",
        NULL
};

static int stream_is_font(AVStream* st)
{
    const AVDictionaryEntry *tag = NULL;
    int n;

    if (st->codecpar->codec_type != AVMEDIA_TYPE_ATTACHMENT)
        return 0;

    tag = av_dict_get(st->metadata, "mimetype", NULL, AV_DICT_MATCH_CASE);

    if (tag) {
        for (n = 0; font_mimetypes[n]; n++) {
            if (av_strcasecmp(font_mimetypes[n], tag->value) == 0)
                return 1;
        }
    }
    return 0;
}

static av_cold int init(AVFilterContext *ctx)
{
    ASS_Context *context = ctx->priv;

    context->library = ass_library_init();
    ass_set_message_cb(context->library, ass_log, NULL);
    ass_set_fonts_dir(context->library, context->fontsdir);
    ass_set_extract_fonts(context->library, 1);

    if (context->filename) {
        AVFormatContext *video = NULL;
        int ret = avformat_open_input(&video, context->filename, NULL, NULL);
        if (ret < 0) {
            av_log(ctx, AV_LOG_ERROR, "Unable to open %s\n", context->filename);
            return ret;
        }
        for (int i = 0; i < video->nb_streams; i++) {
            const AVDictionaryEntry *tag;
            AVStream *st = video->streams[i];
            if (!stream_is_font(st)) continue;
            tag = av_dict_get(st->metadata, "filename", NULL, AV_DICT_MATCH_CASE);
            if (!tag) continue;
            av_log(NULL, AV_LOG_DEBUG, "Loading attached font: %s\n", tag->value);
            ass_add_font(context->library, tag->value,
                 (char *)st->codecpar->extradata,
                 st->codecpar->extradata_size);
        }
        avformat_close_input(&video);
    }

    context->renderer = ass_renderer_init(context->library);
    ass_set_pixel_aspect(context->renderer, 1);
    ass_set_shaper(context->renderer, 0);
    ass_set_fonts(context->renderer, NULL, NULL, 1, NULL, 1);

    context->track = NULL;              // Will be init'd in filter_frame
    context->ass_split_context = NULL;  //

    init_palettizecontext(&context->palettize_context);

    ff_draw_init(&context->draw_context, AV_PIX_FMT_BGRA, FF_DRAW_PROCESS_ALPHA);

    return 0;
}

static int config_input(AVFilterLink *inlink)
{
    const AVFilterContext *ctx = inlink->dst;
    ASS_Context *context = ctx->priv;
    if (!context->size.width)  context->size.width  = inlink->w;
    if (!context->size.height) context->size.height = inlink->h;
    if (!context->size.height || !context->size.width) {
        av_log(NULL, AV_LOG_ERROR, "A positive height and width are required to render subtitles\n");
        return AVERROR_EXIT;
    }
    ass_set_frame_size(context->renderer, context->size.width, context->size.height);
    ass_set_storage_size(context->renderer, context->size.width, context->size.height);

    return 0;
}

static int config_output(AVFilterLink *outlink)
{
    AVFilterContext *ctx = outlink->src;
    ASS_Context *context = ctx->priv;

    outlink->w = context->size.width;
    outlink->h = context->size.height;

    return 0;
}

static int filter_frame(AVFilterLink *inlink, AVFrame *frame)
{
    const AVFilterContext *ctx = inlink->dst;
    ASS_Context *context = ctx->priv;
    AVFilterLink *outlink = ctx->outputs[0];
    if (!context->track) {      // Init track
        if (frame->subtitle_header)
            context->track = ass_read_memory(context->library,
                (char *) frame->subtitle_header->data,
                frame->subtitle_header->size, NULL);
        else context->track = ass_new_track(context->library);
        if (!frame->subtitle_header) context->ass_split_context = avpriv_ass_split("");
        else context->ass_split_context = avpriv_ass_split((char *) frame->subtitle_header->data);
    }

    for (int r = 0; r < frame->num_subtitle_areas; r++)
    {
        ASS_Image *image;
        ASSDialog *dialog;
        ASS_Track *track = context->track;
        AVSubtitleArea *area = frame->subtitle_areas[r];
        if (area->type == SUBTITLE_BITMAP) continue;

        dialog = avpriv_ass_split_dialog(context->ass_split_context, area->ass);
        if (track->n_events > 0)
            ass_free_event(track, 0);
        track->n_events = 0;
        ass_alloc_event(track);
        track->n_events = track->max_events = 1;
        track->events[0].Start = frame->subtitle_timing.start_pts / (AV_TIME_BASE / 1000);
        track->events[0].Duration = frame->subtitle_timing.duration / (AV_TIME_BASE / 1000);
        track->events[0].Effect = strdup(dialog->effect);
        track->events[0].Layer = dialog->layer;
        track->events[0].MarginL = dialog->margin_l;
        track->events[0].MarginR = dialog->margin_r;
        track->events[0].MarginV = dialog->margin_v;
        track->events[0].Name = strdup(dialog->name);
        track->events[0].Text = strdup(dialog->text);
        track->events[0].ReadOrder = dialog->readorder;
        track->events[0].Style = 0;
        for (int style = 0; style < track->n_styles; style++)
            if (!strcmp(track->styles[style].Name, dialog->style))
                track->events[0].Style = style;
        track->events[0].render_priv = NULL;
        avpriv_ass_free_dialog(&dialog);

        image = ass_render_frame(context->renderer, track,
            track->events[0].Start + track->events[0].Duration / 2, NULL);
        if (image == NULL) {
            av_log(NULL, AV_LOG_WARNING, "failed to render ass: %s\n", area->ass);
            return 0;
        }
        ass_image_to_area_palletization(context, image, area);
        area->type = SUBTITLE_BITMAP;

        av_log(NULL, AV_LOG_DEBUG, "successfully rendered ass: %s\n", area->ass);
    }

    return ff_filter_frame(outlink, frame);
}

static av_cold void uninit(AVFilterContext *ctx)
{
    ASS_Context *context = ctx->priv;
    free_palettizecontext(&context->palettize_context);
    avpriv_ass_split_free(context->ass_split_context);

    if (context->track)
        ass_free_track(context->track);
    if (context->renderer)
        ass_renderer_done(context->renderer);
    if (context->library)
        ass_library_done(context->library);

    context->track = NULL;
    context->renderer = NULL;
    context->library = NULL;
}

// Copied from sf_graphicsub2text, with formats swapped
static int query_formats(AVFilterContext *ctx)
{
    AVFilterFormats *formats, *formats2;
    AVFilterLink *inlink = ctx->inputs[0];
    AVFilterLink *outlink = ctx->outputs[0];
    static const enum AVSubtitleType in_fmts[] = { AV_SUBTITLE_FMT_ASS, AV_SUBTITLE_FMT_NONE };
    static const enum AVSubtitleType out_fmts[] = { AV_SUBTITLE_FMT_BITMAP, AV_SUBTITLE_FMT_NONE };
    int ret;

    /* set input format */
    formats = ff_make_format_list(in_fmts);
    if ((ret = ff_formats_ref(formats, &inlink->outcfg.formats)) < 0)
        return ret;

    /* set output format */
    formats2 = ff_make_format_list(out_fmts);
    if ((ret = ff_formats_ref(formats2, &outlink->incfg.formats)) < 0)
        return ret;

    return 0;
}

static void palettize_image(PalettizeContext *const s,
                           const int w, const int h,
                           uint8_t *src_data, const int src_linesize,
                           uint8_t *dst_data, const int dst_linesize,
                           uint32_t *dst_pal, const int num_colors)
{
    const int codeword_length = w * h;

    /* Re-Initialize */
    s->codeword = av_realloc_f(s->codeword, codeword_length, 4 * sizeof(*s->codeword));
    s->codeword_closest_codebook_idxs = av_realloc_f(
        s->codeword_closest_codebook_idxs, codeword_length,
        sizeof(*s->codeword_closest_codebook_idxs));
    s->codebook = av_realloc_f(s->codebook, num_colors, 4 * sizeof(*s->codebook));

    /* build the codeword */
    for (int k = 0, i = 0; i < h; i++)
        for (int j = 0; j < w; j++) {
            s->codeword[k++] = src_data[i * src_linesize + j * 4 + s->b_idx];
            s->codeword[k++] = src_data[i * src_linesize + j * 4 + s->g_idx];
            s->codeword[k++] = src_data[i * src_linesize + j * 4 + s->r_idx];
            s->codeword[k++] = src_data[i * src_linesize + j * 4 + s->a_idx];
        }

    /* compute the codebook */
    avpriv_elbg_do(&s->elbg, s->codeword, 4,
        codeword_length, s->codebook, num_colors, 1,
        s->codeword_closest_codebook_idxs, &s->lfg, 0);

    /* Write Palette */
    for (int i = 0; i < num_colors; i++)
        dst_pal[i] = (s->codebook[i*4+3] << 24) |
                     (s->codebook[i*4+2] << 16) |
                     (s->codebook[i*4+1] <<  8) |
                     (s->codebook[i*4+0] <<  0);

    /* Write Image */
    for (int k = 0, i = 0; i < h; i++)
        for (int j = 0; j < w; j++)
            dst_data[i * dst_linesize + j] =
                s->codeword_closest_codebook_idxs[k++];
}

#define OFFSET(x) offsetof(ASS_Context, x)
#define FLAGS (AV_OPT_FLAG_SUBTITLE_PARAM|AV_OPT_FLAG_FILTERING_PARAM)

static const AVOption text2graphicsub_options[] = {
        { "s",          "output size",                          OFFSET(size),       AV_OPT_TYPE_IMAGE_SIZE,                         .default_val.str = NULL, .flags = FLAGS },
        { "size",       "output size",                          OFFSET(size),       AV_OPT_TYPE_IMAGE_SIZE,                         .default_val.str = NULL, .flags = FLAGS },
        { "n",          "number of output colors",              OFFSET(num_colors), AV_OPT_TYPE_INT,        .min = 2, .max = 256,   .default_val.i64 = 256,  .flags = FLAGS },
        { "num_colors", "number of output colors",              OFFSET(num_colors), AV_OPT_TYPE_INT,        .min = 2, .max = 256,   .default_val.i64 = 256,  .flags = FLAGS },
        { "f",          "media file from which to load fonts",  OFFSET(filename),   AV_OPT_TYPE_STRING,                             .default_val.str = NULL, .flags = FLAGS },
        { "filename",   "media file from which to load fonts",  OFFSET(filename),   AV_OPT_TYPE_STRING,                             .default_val.str = NULL, .flags = FLAGS },
        { "fd",         "fonts directory",                      OFFSET(fontsdir),   AV_OPT_TYPE_STRING,                             .default_val.str = NULL, .flags = FLAGS },
        { "fontsdir",   "fonts directory",                      OFFSET(fontsdir),   AV_OPT_TYPE_STRING,                             .default_val.str = NULL, .flags = FLAGS },
        { .name =  NULL }
};

static const AVClass text2graphicsub_class = {
        .class_name       = "text2graphicsub",
        .item_name        = av_default_item_name,
        .option           = text2graphicsub_options,
        .version          = LIBAVUTIL_VERSION_INT,
        .category         = AV_CLASS_CATEGORY_FILTER,
};

static const AVFilterPad inputs[] = {
    {
        .name         = "default",
        .type         = AVMEDIA_TYPE_SUBTITLE,
        .filter_frame = filter_frame,
        .config_props = config_input,
    },
};

static const AVFilterPad outputs[] = {
    {
        .name         = "default",
        .type         = AVMEDIA_TYPE_SUBTITLE,
        .config_props = config_output,
    },
};

const AVFilter ff_sf_text2graphicsub = {
        .name            = "text2graphicsub",
        .description     = NULL_IF_CONFIG_SMALL("Convert text subtitles to bitmap subtitles."),
        .init            = init,
        .uninit          = uninit,
        .priv_size       = sizeof(ASS_Context),
        .priv_class      = &text2graphicsub_class,
        FILTER_INPUTS(inputs),
        FILTER_OUTPUTS(outputs),
        FILTER_QUERY_FUNC(query_formats)
};
```

Reply:
Ok, so you mean just replace the prototypes with the definitions. What's wrong with having prototypes?

Reply:
> Ok, so you mean just replace the prototypes with the definitions. What's wrong with having prototypes?

I can't even tell you. No idea. I'm just parroting the things they say and want. In this regard, there was just a recent case:

https://www.mail-archive.com/ffmpeg-devel@ffmpeg.org/msg120417.html